src/ray/gcs/redis_client.cc

```diff
@@ -120,7 +121,41 @@ static void GetRedisShards(redisContext *context,
   freeReplyObject(reply);
 }
 
-RedisClient::RedisClient(const RedisClientOptions &options) : options_(options) {}
+RedisClient::RedisClient(const RedisClientOptions &options) : options_(options) {
+  instrumented_io_context io_context;
+  auto tmp_context = std::make_shared<RedisContext>(io_context);
+  RAY_CHECK_OK(tmp_context->Connect(options_.server_ip_,
+                                    options_.server_port_,
+                                    /*sharding=*/options_.enable_sharding_conn_,
+                                    /*password=*/options_.password_,
+                                    /*enable_ssl=*/options_.enable_ssl_));
+  auto reply = tmp_context->RunArgvSync(std::vector<std::string>{"INFO", "REPLICATION"});
+  RAY_CHECK(reply && !reply->IsNil()) << "Failed to get Redis replication info";
+  auto replication_info = reply->ReadAsString();
+  auto parts = absl::StrSplit(replication_info, "\r\n");
+
+  for (auto &part : parts) {
+    if (part.empty() || part[0] == '#') {
+      continue;
+    }
+    std::vector<std::string> kv = absl::StrSplit(part, ":");
+    RAY_CHECK(kv.size() == 2);
+    if (kv[0] == "role" && kv[1] == "master") {
+      leader_ip_ = options_.server_ip_;
+      leader_port_ = options_.server_port_;
+      break;
+    }
+
+    if (kv[0] == "master_host") {
+      leader_ip_ = kv[1];
+    }
+    if (kv[0] == "master_port") {
+      leader_port_ = std::stoi(kv[1]);
+    }
+  }
+  RAY_LOG(INFO) << "Find redis leader: " << leader_ip_ << ":" << leader_port_;
```

Comment:
in redis, would a leader ever demote to a follower? if that's the case, would write to follower fail?

Reply:
I thought about this and actually tested it manually. In this case, writing will fail and then GCS crash.
I will test test for this.